Lib/asyncio/streams.py

```diff
@@ -307,6 +302,30 @@ def __del__(self):
             closed.exception()
 
 
+class StreamReaderProtocol(BaseStreamReaderProtocol, protocols.Protocol):
+    def data_received(self, data):
+        reader = self._stream_reader
+        if reader is not None:
+            reader.feed_data(data)
+
+
+class StreamReaderBufferedProtocol(BaseStreamReaderProtocol, protocols.BufferedProtocol):
```

Comment:
Sorry, I don't follow.
Where is `StreamReaderBufferedProtocol` class used (except tests)?

Reply:
It isn't used right now, as there is no way to include this class without changing or adding to the API.
It will be useful in the future though, once we do make API changes.

Reply:
Please elaborate.
From my understanding, the PR contains dead code only. Despite the code correctness, it is not used by the implementation of the streams.
What future API changes are you talking about? Can we apply these changes as a part of this pull request?

I like this PR, it makes a value if the buffered protocol is used. But if this protocol is not used -- why we need the PR at all?

I can imagine a little different approach: don't extract `BaseStreamReaderProtocol` but inherit `StreamReaderProtocol` from both `Protocol` and `BufferedProtocol`, implementing all `data_received()`, `get_buffer()` and `update_buffer()` in the same class.
asyncio selects the best approach currently: it uses the buffered protocol version if the protocol is an instance of `BufferedProtocol` with fallback to `data_received()` otherwise. We can utilize this facility.

What do you think?

Reply:
The problem with this approach is that the `BufferedProtocol` allocates a 64k byte buffer, so on servers with a lot of connections, memory will start running out quickly as each stream now holds a 64k buffer.

For example on a server with 10k clients, python will hold ~640mb of memory.

This is why we should probably make it a custom choice of when to use the new `BufferedProtocol` *or* creating some kind of a protocol pool of a small size, so that the first few streams will be faster.

But this means that we either change / add API or choose a set number defaulted to the protocol pool.

Btw, doing the buffered protocol pool approach is not really trivial as I don't currently are how we know when a `BufferedProtoco`l has gone out of the pool (maybe I can think of something if you tell me that it's worth it, I'll be happy to implement something like that)

Reply:
We already have `limit` argument to control *output* buffer.
We can reuse the same number for controlling *input* buffer size as well.
In practice, I very doubt if people need different values for that.

I can imagine a variadic input buffer size starting from some default value and growing up to a limit but I doubt if we need it really. A simple implementation can be good enough.

A pool of buffers is an interesting idea but, again, the feature can be added later.

Thoughts?

Reply:
> We already have limit argument to control output buffer.
> We can reuse the same number for controlling input buffer size as well.
> In practice, I very doubt if people need different values for that.

I don't really understand what you mean by using the ``limit`` argument as an *input* buffer size, as ``StreamReader``'s ``limit`` argument actually controls the buffer limit of what the ``read`` apis (such as ``read_exactly`` *or* ``read_line``) use to get the buffer from, so really this argument controls the *input* buffer size.

> I can imagine a variadic input buffer size starting from some default value and growing up to a limit but I doubt if we need it really. A simple implementation can be good enough.

I believe what you mean is already implemented in ``StreamReader`` but I am not sure.
Also I am against reallocating the buffer as it would remove all benefits of ``BufferedProtocol``.

> A pool of buffers is an interesting idea but, again, the feature can be added later.

I think this is the best solution, though we need to think if we want that feature to be in this PR or on a different one.

Reply:
I reckon that the default buffer size is not necessarily decided by `limit` argument. We might start with a relatively small buffer, e.g. `io.DEFAULT_BUFFER_SIZE` or less, then expand and shrink buffer on demand.

Additionally, it appears that `feed_data` inflicts extra copy, so is it viable to directly write the buffer in `BaseStreamReaderProtocol` instead of the buffer in `StreamReaderBufferedProtocol`, or more aggresively utilize efficient and mature [`io.BufferedIOBase`](https://docs.python.org/3/library/io.html#io.BufferedIOBase) to implement `StreamReaderBufferedProtocol`?